server/cherry_pick.go

```diff
@@ -23,6 +23,47 @@ import (
 	"github.com/pkg/errors"
 )
 
+const (
+	cherryPickScheduledMsg = "Cherry pick is scheduled."
+	tooManyCherryPickMsg   = "There are too many cherry picking requests. Please do this manually or try again later."
+)
+
+type cherryPickRequest struct {
+	pr        *model.PullRequest
+	version   string
+	milestone int
+}
+
+func (s *Server) listenCherryPickRequests() {
+	for {
+		select {
+		case job := <-s.cherryPickRequests:
+			ctx, cancel := context.WithTimeout(context.Background(), defaultRequestTimeout*time.Second)
+			defer cancel()
+			pr := job.pr
+			cmdOut, err := s.doCherryPick(ctx, strings.TrimSpace(job.version), &job.milestone, pr)
+			if err != nil {
+				msg := fmt.Sprintf("Error trying doing the automated Cherry picking. Please do this manually\n\n```\n%s\n```\n", cmdOut)
+				s.sendGitHubComment(ctx, pr.RepoOwner, pr.RepoName, pr.Number, msg)
+				mlog.Error("Error while cherry picking", mlog.Err(err))
+			}
+		case <-s.stopChan:
+			close(s.cherryPickRequests)
+			s.drainCherryPickTasks()
```

Comment:
Not sure if we need such a complicated way of doing things. We can just range over the channel, and then close it and send another signal either through waitgroup or another channel to the parent goroutine. I would also prefer the channels to be prefixed with the name `cherryPick` so that it's clear it's for cherry picking and not meant to be used at a server level.

```go
func (s *Server) listenCherryPickRequests() {
 defer func() {
   close(s.cherryPickStoppedChan)
   // or alternatively use a waitgroup. Either is fine.
 }()

 for job := range s.cherryPickRequests {
  // handle job
 }
}
```

```go
func (s *Server) stopCherryPickListener() {
 close(c.cherryPickRequests)
 <-s.cherryPickStoppedChan
}
```
Then we just have a single place to do everything. And the pending jobs in the queue don't get canceled when the server is shutting down.

Reply:
That was my initial thought about this but in that case we need to wait every cherry-pick request to complete. If there are too many requests queued up, server shutdown will take too much time, I did this to be able to cancel remaining requests.

Reply:
I see. But in your case, we are cancelling everything the moment we get a stop request. I think it is reasonable for the server to have to wait for a while to flush out pending cherry-pick requests, within a timeout of course.

I see 2 options here:

- How much of a time on average does a cherry-pick take? Depending on that, we might want to reduce the buffer size. So then server shutdown never takes too long.
- We can add additional timeout mechanism in the `stopCherryPickListener` method to exit the server if it has to wait beyond a certain time. But then those "cancelling requests" comment doesn't get sent out.

I think reducing the buffer size is an automatic way of ensuring the server shuts down in a reasonable time and also ensuring that the user gets a response.

Reply:
I think we can prioritize informing user about the outcome if it is canceled/failed/succeeded. I think the answer of "how often do we restart the server" can make more sense to give a decision about this. Also, cherry-pick shouldn't take too much but it depends to too many components. Sometimes it takes too long to do `gh checkout pr n`. So the option 1 makes more sense to me.

From my understanding we have to decide to give-up cancelling requests over code complexity. Anyway, we can reduce the queue size at the beginning and wait every cherry-pick request to complete. After some usage we may have a better picture what path should we use.

Reply:
> From my understanding we have to decide to give-up cancelling requests over code complexity. 

I don't mind the code complexity really. IMO, this is what should be done: while closing the server, we send a signal to the goroutine, and wait for x seconds for it to return. In those x seconds, it can flush out the queue. And when the timeout is over, the remaining items need to be canceled and sent to the user.

This is what I would want in an ideal world. This has almost everything in your first solution, but instead of outright cancelling, it waits for a certain period. Now I think implementing this can get quite complicated. But I am 0/5 if you see a reasonable way to do this.

If not, I think our current solution is good enough :) So yeah, we are compromising on code complexity for our "ideal" solution. But if the "ideal" solution is doable, then I'm all for it.